File: 500/525 - Contiguous Array.java
import java.util.HashMap;

class Solution {
    public int findMaxLength(int[] nums) {
        int ret = 0, i, count = 0;
        HashMap<Integer, Integer> hm = new HashMap<Integer, Integer>();
        
        hm.put(0, -1);
        for (i = 0; i < nums.length; ++i) {
            if (nums[i] == 1)
                ++count;
            else
                --count;
            
            if (hm.containsKey(count))
                ret = Math.max(ret, i - hm.get(count));
            else
                hm.put(count, i);
        }
        
        return ret;
    }
}

File: csharp/525.contiguous-array.cs
/*
 * @lc app=leetcode id=525 lang=csharp
 *
 * [525] Contiguous Array
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {
    public int FindMaxLength(int[] nums) {
        int ret = 0, n = nums.Length, curr = 0;
        IDictionary<int, int> map = new Dictionary<int, int>();
        map.Add(0, -1);
        for (int i = 0; i < n; ++i) {
            curr += nums[i] == 0 ? 1 : -1;
            int j;
            if (map.TryGetValue(curr, out j)) {
                ret = Math.Max(ret, i - j);
            } else {
                map[curr] = i;
            }
        }
        return ret;
    }
}
// @lc code=end

